Modules/_threadmodule.c

```diff
@@ -42,24 +42,74 @@ get_thread_state(PyObject *module)
 
 // _ThreadHandle type
 
+// Handles transition from RUNNING to one of JOINED, DETACHED, or INVALID (post
+// fork).
+typedef enum {
+    THREAD_HANDLE_RUNNING,
+    THREAD_HANDLE_JOINED,
+    THREAD_HANDLE_DETACHED,
+    THREAD_HANDLE_INVALID,
+} ThreadHandleState;
+
+// A handle around an OS thread.
+//
+// The OS thread is either joined or detached after the handle is destroyed.
+//
+// Joining the handle is idempotent; the underlying OS thread is joined or
+// detached only once. Concurrent join operations are serialized until it is
+// their turn to execute or an earlier operation completes successfully. Once a
+// join has completed successfully all future joins complete immediately.
 typedef struct {
     PyObject_HEAD
     struct llist_node node;  // linked list node (see _pythread_runtime_state)
+
+    // The `ident` and `handle` fields are immutable once the object is visible
+    // to threads other than its creator, thus they do not need to be accessed
+    // atomically.
     PyThread_ident_t ident;
     PyThread_handle_t handle;
-    char joinable;
+
+    // Holds a value from the `ThreadHandleState` enum.
+    int state;
+
+    // Set immediately before `thread_run` returns to indicate that the OS
+    // thread is about to exit.
+    _PyEventRc *thread_is_exiting;
+
+    // Serializes calls to `join`.
+    _PyOnceFlag once;
 } ThreadHandleObject;
 
+static inline int
+get_thread_handle_state(ThreadHandleObject *handle)
+{
+    return _Py_atomic_load_int_relaxed(&handle->state);
```

Comment:
I don't know if this is logically ok, but a good rule of thumb is that the "relaxed" variants should only really be used for performance-critical code, which is not the case here.

Reply:
I think we're OK here. The rationale being that we only need to access `state` atomically and don't thus don't need the ordering guarantees provided by the stronger memory orders. That said, I considered wrapping `ThreadHandle_join` in a critical section, but it feels a little heavy-handed.

I think it's also worth being painfully obvious that this is only relevant for the free-threaded build. There's no risk to correctness in the default build.

Reply:
I wouldn't really rely on "I think we're ok" given that we're talking about potentially hairy and difficult to reproduce issues (such as ordering of loads and stores on different CPUs).

Since this isn't performance-critical all, my vote vote is strongly in favor of the "normal" atomic API (i.e. sequentially consistent).

Reply:
> Since this isn't performance-critical all, my vote vote is strongly in favor of the "normal" atomic API (i.e. sequentially consistent).

Good point!